bson/src/main/org/bson/codecs/NumberCodecHelper.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs;
+
+import org.bson.BsonInvalidOperationException;
+import org.bson.BsonReader;
+import org.bson.BsonType;
+
+import static java.lang.String.format;
+import static org.bson.assertions.Assertions.isTrueArgument;
+
+final class NumberCodecHelper {
+
+    @SuppressWarnings("unchecked")
+    static <T extends Number> T decodeNumber(final BsonReader reader, final Class<T> clazz) {
```

Comment:
I think it's worth breaking up this method into three specialized ones, one for each of the three supported numeric types.  This will avoid warning suppression, unnecessary primitive boxing/unboxing, redundant casts, and checks that are unnecessary if you know the expected type.  As an example for double, I'd prefer something more like this:

    static double decodeDouble(final BsonReader reader) {
        double doubleValue;
        BsonType bsonType = reader.getCurrentBsonType();
        switch (bsonType) {
            case INT32:
                doubleValue = reader.readInt32();
                break;
            case INT64:
                long longValue = reader.readInt64();
                doubleValue = longValue;
                if (longValue != (long) doubleValue) {
                    throw invalidConversion(Double.class, longValue);
                }
                break;
            case DOUBLE:
                doubleValue = reader.readDouble();
                break;
            default:
                throw new BsonInvalidOperationException(format("Invalid numeric type, found: %s", bsonType));
        }
        return doubleValue;
    }